### PR TITLE
Support alert routing based on labels

### PR DESCRIPTION
**Pull request by joeyorlando**
# What this PR does

This PR adds support for routing alerts based on labels.
https://www.loom.com/share/4401de6e3c4945d5b8961fe43ee373c9

Additionally:
- improve the typing around the `get_object` method that is inherited by [`PublicPrimaryKeyMixin.get_object`](https://github.com/grafana/oncall/blob/dev/engine/common/api_helpers/mixins.py#L153) in most of our models. `PublicPrimaryKeyMixin` is generic, so it can be more strongly typed when it is being subclassed, which results in better typing of the `get_object` method in child classes
  - I decided to do this because I started looking into this task via the [`AlertReceiveChannelView.send_demo_alert` method/endpoint](https://github.com/grafana/oncall/blob/dev/engine/apps/api/views/alert_receive_channel.py#L242). Within that method, `instance` is not typed because the inherited `get_object` method is not typed.. I digress 😄 
- improve typing around `Alert.create` and `apps.integrations.tasks.create_alert` functions
- make `Alert.render_group_data` more DRY by extracting some logic out into `Alert._apply_jinja_template_to_alert_payload_and_labels`
- deduplicate the logic of `value.strip().lower() in ["1", "true", "ok"]` into a shared function, `common.jinja_templater.apply_jinja_template.templated_value_is_truthy`

Closes https://github.com/grafana/oncall-private/issues/2490

## Checklist

- [x] Unit, integration, and e2e (if applicable) tests updated
- [x] `CHANGELOG.md` updated (or `pr:no changelog` PR label added if not required)
- [x] Documentation added (or `pr:no public docs` PR label added if not required) (will be done in #3762)

### CODE REVIEW

File: engine/apps/user_management/models/user.py [hunk 1]
```diff
@@ -272,10 +279,6 @@ def clear_phone_numbers(self) -> None:
         self._verified_phone_number = None
         self.save(update_fields=["unverified_phone_number", "_verified_phone_number"])
 
-    # TODO: move to telegram app
-    def is_telegram_connected(self):
-        return hasattr(self, "telegram_connection")
-
```
Comment (joeyorlando) [inline engine/apps/user_management/models/user.py:278]: `mypy` [caught a bug](https://github.com/grafana/oncall/actions/runs/7700348181/job/20983996764?pr=3778#step:4:535) on CI:
```
apps/api/views/user.py:603: error: Function "is_telegram_connected" could always be true in boolean context  [truthy-function]
Found 1 error in 1 file (checked 670 source files)
```

[Here](https://github.com/grafana/oncall/blob/dev/engine/apps/api/views/user.py#L603) we're doing:
```python3
if not user.is_telegram_connected:
            return Response(status=status.HTTP_400_BAD_REQUEST)
```
in its current state, `User.is_telegram_connected` is a method, not a property so `not user.is_telegram_connected` is not evaluating as we likely originally intended...

There's only one reference to this method (now a property), so this should be fine refactoring as such.

File: engine/apps/alerts/models/channel_filter.py
Comment (joeyorlando) [inline engine/apps/alerts/models/channel_filter.py:1]: one of the main changes

File: engine/apps/alerts/models/alert.py [hunk 2]
```diff
@@ -110,7 +118,7 @@ def create(
         )
 
         if group_created:
-            assign_labels(group, alert_receive_channel, raw_request_data)
+            assign_labels(group, alert_receive_channel, parsed_labels)
```
Comment (joeyorlando) [inline engine/apps/alerts/models/alert.py:121]: some of the main changes

File: engine/apps/labels/alert_group_labels.py
Comment (joeyorlando) [inline engine/apps/labels/alert_group_labels.py:1]: main changes

File: engine/apps/alerts/models/channel_filter.py [hunk 3]
```diff
@@ -126,29 +132,30 @@ def select_filter(cls, alert_receive_channel, raw_request_data, force_route_id=N
                 )
                 pass
 
-        filters = cls.objects.filter(alert_receive_channel=alert_receive_channel)
+        channel_filters = cls.objects.filter(alert_receive_channel=alert_receive_channel)
 
-        satisfied_filter = None
-        for _filter in filters:
-            if satisfied_filter is None and _filter.is_satisfying(raw_request_data):
-                satisfied_filter = _filter
```
Comment (joeyorlando) [inline engine/apps/alerts/models/channel_filter.py:134]: I didn't see a reason to continue looping over `filters` if `satisifed_filter` is already set and being returned just below

File: engine/apps/alerts/models/alert.py [hunk 4]
```diff
@@ -79,28 +81,34 @@ def get_integration_optimization_hash(self):
...
+    ) -> "Alert":
         """
         Creates an alert and a group if needed.
         """
         # This import is here to avoid circular imports
         from apps.alerts.models import AlertGroup, AlertGroupLogRecord, AlertReceiveChannel, ChannelFilter
 
+        parsed_labels = gather_labels_from_alert_receive_channel_and_raw_request_data(
+            alert_receive_channel, raw_request_data
+        )
```
Comment (iskhakov) [inline engine/apps/alerts/models/alert.py:105]: render_group_data() on line 106 is parsing group_id, etc out of alert and stores in AlertGroup.GroupData, so it makes sense to parse labels there as well

File: engine/apps/alerts/tests/test_channel_filter.py [hunk 5]
```diff
@@ -91,3 +91,34 @@ def test_channel_filter_select_filter_jinja2(make_organization, make_alert_recei
...
+
+@pytest.mark.django_db
+@pytest.mark.parametrize(
+    "filtering_term,should_match",
+    [
+        ('{{ "foo" in labels.keys() }}', True),
+        ('{{ "bar" in labels.values() }}', True),
+        ('{{ "bar" in labels.values() or payload["value"] == 5 }}', True),
+        ('{{ "baz" in labels.values() }}', False),
+    ],
```
Comment (iskhakov) [inline engine/apps/alerts/tests/test_channel_filter.py:104]: Let's check the cases `{{ labels.foo=="bar"}}`, `{{ labels.foo=="bar" and labels.bar=="baz" }}`, `{{ labels.foo=="bar" or labels.bar=="baz" }}`

I think thats what will be generated by the UI template builder

Comment (joeyorlando) [reply engine/apps/alerts/tests/test_channel_filter.py:104]: good call 👍

File: engine/apps/alerts/models/alert.py [hunk 6]
```diff
@@ -240,30 +279,31 @@ def render_group_data(cls, alert_receive_channel, raw_request_data, is_demo=Fals
...
-                is_acknowledge_signal = apply_jinja_template(acknowledge_condition_template, payload=raw_request_data)
-            except (JinjaTemplateError, JinjaTemplateWarning) as e:
-                logger.warning(
-                    f"acknowledge_condition_template error on channel={alert_receive_channel.public_primary_key}: {e.fallback_message}"
-                )
-
-            if isinstance(is_acknowledge_signal, str):
-                is_acknowledge_signal = is_acknowledge_signal.strip().lower() in ["1", "true", "ok"]
-            else:
-                is_acknowledge_signal = False
```
Comment (joeyorlando) [inline engine/apps/alerts/models/alert.py:266]: this was simply deduplicating a lot of copy/paste into `Alert._apply_jinja_template_to_alert_payload_and_labels`

File: engine/apps/alerts/models/channel_filter.py [hunk 7]
```diff
@@ -126,29 +137,36 @@ def select_filter(cls, alert_receive_channel, raw_request_data, force_route_id=N
...
-
-    def check_filter(self, value):
+    def check_filter(
+        self, raw_request_data: "Alert.RawRequestData", alert_labels: typing.Optional["Labels"] = None
+    ) -> bool:
         if self.filtering_term_type == ChannelFilter.FILTERING_TERM_TYPE_JINJA2:
             try:
-                is_matching = apply_jinja_template(self.filtering_term, payload=value)
-                return is_matching.strip().lower() in ["1", "true", "ok"]
+                return templated_value_is_truthy(
```
Comment (joeyorlando) [inline engine/apps/alerts/models/channel_filter.py:157]: I decided to move the duplicated logic of `<string>.strip().lower() in ["1", "true", "ok"]` (we also duplicated this 3x in the `Alert` model) to [`common.jinja_templater.apply_jinja_template.templated_value_is_truthy`](https://github.com/grafana/oncall/pull/3778/files#diff-e14099bf4a01d2763b4c119bb91a0ab0fba257487ec1157b7aac5867afdf57d0R61-R64)

File: engine/apps/alerts/models/alert.py (hunk 4 above)
Comment (joeyorlando) [reply engine/apps/alerts/models/alert.py:110]: good call, done 👍

**Pull request merged**